lib/rom/command.rb

```diff
@@ -20,36 +20,31 @@ module ROM
   #
   # @private
   class Command
+    extend Dry::Initializer::Mixin
     include Dry::Equalizer(:relation, :options)
     include Commands
     include Pipeline::Operator
 
     extend ClassMacros
     extend ClassInterface
 
-    include Options
-
     defines :adapter, :relation, :result, :input, :register_as, :restrictable
 
-    option :type, allow: [:create, :update, :delete]
-    option :source, reader: true
-    option :result, reader: true, allow: [:one, :many]
+    # @attr_reader [Relation] relation The command's relation
+    param :relation
+
+    CommandType = Dry::Types['strict.symbol'].enum(:create, :update, :delete)
+    Result = Dry::Types['strict.symbol'].enum(:one, :many)
+
+    option :type, type: CommandType, optional: true
+    option :source, reader: true, optional: true, default: -> c { c.relation }
+    option :result, reader: true, type: Result
     option :input, reader: true
-    option :curry_args, type: Array, reader: true, default: EMPTY_ARRAY
+    option :curry_args, reader: true, default: -> _ { EMPTY_ARRAY }
```

Comment:
I'd not go with a deep analysis, because no one stops you from doing like this
```ruby
cache = {}
param :cache, default: proc { cache }
```
From my point of view checking `.frozen?` is just enough to be sure that the user knows what he or she is doing. 
Just my thoughts, I don't insist on any changes :)